src/content/en/fundamentals/engage-and-retain/push-notifications/subscribing-a-user.md

```diff
@@ -0,0 +1,396 @@
+push is supported with two simple checks.
+
+1. Check the *serviceWorker* API on the *navigator*.
+1. Check *PushManager* on  the *window*.
+
+      if (!('serviceWorker' in navigator)) {
+        // Service Worker isn't supported on this browser, disable or hide UI.
+        return;
+      }
+
+      if (!('PushManager' in window)) {
+        // Push isn't supported on this browser, disable or hide UI.
+        return;
+      }
+
+While browser support is growing quickly for both service worker and
+push messaging support, it's always a good idea to feature detect for both and
+[progressively enhance](https://en.wikipedia.org/wiki/Progressive_enhancement).
+
+## Register a Service Worker
+
+With the feature detect we know that both service workers and Push is supported. The next step
+is to "register" our service worker.
+
+When we register a service worker, we are telling the browser where our service worker file is.
+The file is still just JavaScript, but the browser will "give it access" to the service worker
+APIs, including push. To be more exact, the browser runs the file in a service worker
+environment.
+
+To register a service worker, call `navigator.serviceWorker.register()`, passing in the path to
+our file. Like so:
+
+    function registerServiceWorker() {
+      return navigator.serviceWorker.register('service-worker.js')
+      .then(function(registration) {
+        console.log('Service worker successfully registered.');
+        return registration;
+      })
+      .catch(function(err) {
+        console.error('Unable to register service worker.', err);
+      });
+    }
+
+This code above tells the browser that we have a service worker file and where its located. In
+this case, the service worker file is at `/service-worker.js`. Behind the scenes the browser
+will take the following steps after calling `register()`:
+
+1. Download the service worker file.
+
+1. Run the JavaScript.
+
+1. If everything ran correctly and there were no errors, the promise returned by `register()`
+will resolve. If there are errors of any kind, the promise will reject.
+
+> If `register()` does reject, double check your JavaScript for typos / errors in Chrome DevTools.
+
+When `register()` does resolve, it returns a `ServiceWorkerRegistration`. We'll use this
+registration to access to the [PushManager
+API](https://developer.mozilla.org/en-US/docs/Web/API/PushManager).
+
+## Requesting Permission
+
+We've registered our service worker and ready to subscribe the user, the next step is to get
+permission from the user to send them push messages.
+
+The API for getting permission is relatively simple, the downside is that
+the API [recently changed from taking a callback to returning a
+Promise](https://developer.mozilla.org/en-US/docs/Web/API/Notification/requestPermission). The
+problem with this, is that we can't tell what version of the API is implemented by the current
+browser, so you have to implement both and handle both.
```

Comment:
Not that I know of and this is largely a mess because the API changed so it's less of a an issue to do with converting a callback to a promise and more to do with the API possibly expecting a callback or returning a promise.